contracts/src/Teleporter/ReceiptQueue.sol

```diff
@@ -8,90 +8,94 @@ pragma solidity 0.8.18;
+    function enqueue(
+        TeleporterMessageReceiptQueue storage queue,
+        TeleporterMessageReceipt memory receipt
+    ) internal {
+        queue.data[queue.last++] = receipt;
     }
 
     /**
      * @dev Removes the oldest open receipt from the queue.
      *
      * Requirements:
-     *
-     * - `msg.sender` must be the owner.
      * - The queue must be non-empty.
      */
-    function dequeue()
-        external
+    function dequeue(
+        TeleporterMessageReceiptQueue storage queue
+    )
+        internal
         returns (TeleporterMessageReceipt memory result)
     {
-        if (msg.sender != owner) {
-            revert Unauthorized();
-        }
-
-        uint256 first_ = first;
+        uint256 first_ = queue.first;
+        if (queue.last == first_) revert EmptyQueue();
+        result = queue.data[first_];
+        delete queue.data[first_];
+        queue.first = first_ + 1;
+    }
 
-        if (last == first_) {
-            revert EmptyQueue(); // empty queue
+    /**
+     * @dev Returns the outstanding receipts for the given chain ID that should be included in the next message sent.
+     */
+    function getOutstandingReceiptsToSend(
+        TeleporterMessageReceiptQueue storage queue
+    )
+        internal
+        returns (TeleporterMessageReceipt[] memory result)
+    {
+        // Get the current outstanding receipts for the given chain ID.
+        // If the queue contract doesn't exist, there are no outstanding receipts to send.
+        uint256 outstandingReceiptsSize = size(queue);
+        if (outstandingReceiptsSize == 0) {
+            return new TeleporterMessageReceipt[](0);
         }
 
-        result = queue[first_];
-
-        delete queue[first_];
-        first = first_ + 1;
+        // Calculate the result size as the minimum of the number of receipts and maximum batch size.
+        uint256 resultSize = _MAXIMUM_RECEIPT_COUNT;
+        if (outstandingReceiptsSize < _MAXIMUM_RECEIPT_COUNT) {
+            resultSize = outstandingReceiptsSize;
+        }
 
-        emit Dequeue(result.receivedMessageID, result.relayerRewardAddress);
+        result = new TeleporterMessageReceipt[](resultSize);
+        for (uint256 i = 0; i < resultSize; ++i) {
+            result[i] = dequeue(queue);
+        }
     }
 
     /**
      * @dev Returns the number of open receipts in the queue.
      */
-    function size() external view returns (uint256) {
-        return last - first;
+    function size(TeleporterMessageReceiptQueue storage queue) internal view returns (uint256) {
+        return queue.last - queue.first;
+    }
+
+    /**
+     * @dev Returns the receipt at the given index in the queue.
+     */
+    function getReceiptAtIndex(TeleporterMessageReceiptQueue storage queue, uint256 index)
+        internal
+        view
+        returns (TeleporterMessageReceipt memory)
+    {
+        return queue.data[index];
```

Comment:
In addition to changing this to treat the index as the index in the queue, could you add a unit test for it? 🙏